File: c++/array/LongestConsecutiveSequence.cpp
class Solution {
public:
    int longestConsecutive(std::vector<int> &num) {
        std::unordered_map<int, bool> dictionary;
        int result = 0;
        for (int i = 0; i < int(num.size()); i ++) {
            dictionary[num[i]] = false;
        }
        
        for (int i = 0; i < int(num.size()); i ++) {
            int curr = 1;
            dictionary[num[i]] = 0;
            int left = num[i] - 1;
            while (dictionary.find(left) != dictionary.end() and !dictionary[left]) {
                dictionary[left] = true;
                curr ++;
                left --;
            }
            int right = num[i] + 1;
            while (dictionary.find(right) != dictionary.end() and !dictionary[right]) {
                dictionary[right] = true;
                curr ++;
                right ++;
            }
            result = std::max(curr, result);
        }
        
        return result;
    }
};

File: archive/python/array/LongestConsecutiveSequence.py
class Solution:
    # @param num, a list of integer
    # @return an integer
    def longestConsecutive(self, num):
        result = 0
        d = {}
        for i in xrange(len(num)):
            d[num[i]] = 1
        
        for i in xrange(len(num)):
            curr = 1
            d[num[i]] = 0
            left = num[i]-1 
            while left in d and d[left] == 1:
                d[left] = 0
                left -= 1
                curr += 1
            right = num[i]+1
            while right in d and d[right] == 1:
                d[right] = 0
                right += 1
                curr += 1
            result = max(curr, result)
            
        return result
